grpc-protocol/src/main/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClient.java

```diff
@@ -162,27 +188,62 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) {
                            return responseFuture;
                        }), ctx.eventLoop());
         }
 
         private static Subscriber<DeframedMessage> singleSubscriber(
-                AggregatedHttpResponse msg, CompletableFuture<HttpResponse> responseFuture) {
+                ClientRequestContext ctx, AggregatedHttpResponse msg, SerializationFormat serializationFormat,
+                CompletableFuture<HttpResponse> responseFuture) {
 
             return new Subscriber<DeframedMessage>() {
+                private HttpData content = HttpData.empty();
+                @Nullable
+                private HttpHeaders trailers;
+
                 @Override
                 public void onSubscribe(Subscription s) {
-                    s.request(1);
+                    // At least 2 requests are required for receiving trailers.
+                    s.request(2);
+                }
+
+                private boolean isTrailers(DeframedMessage message) {
+                   return message.type() >> 7 == 1;
                 }
 
                 @Override
-                public void onNext(DeframedMessage unframed) {
-                    final ByteBuf buf = unframed.buf();
+                public void onNext(DeframedMessage message) {
+                    if (UnaryGrpcSerializationFormats.isGrpcWeb(serializationFormat) && isTrailers(message)) {
+                        final ByteBuf buf;
+                        try {
+                            buf = InternalGrpcWebUtil.messageBuf(message, ctx.alloc());
+                        } catch (Throwable t) {
+                            responseFuture.completeExceptionally(t);
+                            return;
+                        }
+                        try {
+                            trailers = InternalGrpcWebUtil.parseGrpcWebTrailers(buf);
+                            if (trailers == null) {
+                                // Malformed trailers.
+                                responseFuture.completeExceptionally(
+                                        new ArmeriaStatusException(StatusCodes.INTERNAL,
+                                                                   serializationFormat.uriText() +
+                                                                   " trailers malformed: " +
+                                                                   buf.toString(StandardCharsets.UTF_8)));
+                            }
+                        } finally {
+                            buf.release();
+                        }
+                        return;
+                    }
+                    final ByteBuf buf = message.buf();
                     // Compression not supported.
                     assert buf != null;
-                    responseFuture.complete(HttpResponse.of(msg.headers(), HttpData.wrap(buf), msg.trailers()));
+                    content = HttpData.wrap(buf);
                 }
 
                 @Override
```

Comment:
ditto. How about releasing `content` when an exception is raised?

Reply:
done, PTAL